### PR TITLE
Add account selection spinner to account settings

### LINKED ISSUE #3747: Move settings import/export to 'general settings' screen

### PR DESCRIPTION
**Pull request by williamvds**
This brings together the first pages of general settings and account settings into a tabbed layout:

generalsettingsredesign
accountsettingsredesign

Fixes #3747 
Roughly follows design from https://github.com/k9mail/k-9/issues/1859#issuecomment-268657013 and https://github.com/k9mail/k-9/issues/1859#issuecomment-270029045

### TODO
- add search action to activity
- fix import and export
  - only currently works when 'Start in Unified Inbox' is enabled
- account deletion (using method introduced by https://github.com/k9mail/k-9/pull/3840)
- summaries for account settings categories
- use account color as the background for the spinner
- add vertical padding to spinner drop-down

### CODE REVIEW

File: app/ui/src/main/java/com/fsck/k9/ui/settings/SettingsActivity.kt [hunk 1]
```diff
@@ -1,108 +1,189 @@
...
+        }
+
+        override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+            clearAccounts()
+
+            val prefs = Preferences.getPreferences(context)
+
+            val viewModel: SettingsViewModel by viewModel()
+            viewModel.accounts.observeNotNull(this) {
+                loadAccounts(prefs.getAccounts())
```
Comment (williamvds) [inline app/ui/src/main/java/com/fsck/k9/ui/settings/SettingsActivity.kt]: I hit an issue when trying to use this observer the proper way, if the default activity is the accounts page, and you
1. Launch app
2. Open settings
3. Go back to accounts activity
4. Delete the first account
5. Re-open settings

The observer runs twice after re-opening
The first time it gives the list before account removal, in the second it has been removed
Is there a better way to prevent or work around this issue?

**Comment by ByteHamster** [comment]
Would be great if the settings search feature would still be supported after this change.

**Comment by williamvds** [comment]
> Would be great if the settings search feature would still be supported after this change.

Oh yeah, forgot to add that to the to-do list

**Comment by cketti** [comment]
I don't like the tabs but the way it is now. Making it easier to switch between different accounts would be good, though. But I think having a dropdown in the toolbar is the way to go here.

**Comment by williamvds** [comment]
Thanks for the feedback
Accounts are listed under the first page of general settings:

redesignsettings2

I've moved the account spinner into the account settings toolbar, aligned to the right

redesignaccountsettings

The spinner launches a new account settings activity and `finish()`es the current one
However, on my device, swapping accounts rapidly often results in a strict mode instance count violation, force closing the activity and launching a new one with the default account.
Apparently that's a fairly common issue - is there a good workaround for it?

I initially tried just replacing the settings fragment, but the fragment kept loading with the initial account uuid, no matter what arguments I `create()`ed it with

**Comment by wiktor-k** [comment]
Okaaay, for me it looks a lot better... I'll give more detailed feedback after I check this on a real device... :+1:

**Comment by wiktor-k** [comment]
William, this looks very, *very* good.

One minor thing... that I don't know how to resolve but the titles and account spinner have a little space to breath. I thought about combining the spinner and screen title into one but I'm not sure if it's a good idea.

How it looks on my device:

dots1

dots2

There were a couple of PR merged recently and thus this PR got some conflicts to resolve... (but you probably see this already ;) ).

Thanks for your work!

**Comment by williamvds** [comment]
Resolved conflicts, moved spinner just below (but combined with) the toolbar:
accountsettingsredesign2

**Comment by Valodim** [comment]
Awesome! Definitely an improvement at this point :+1:

**Comment by wiktor-k** [comment]
Yep, looks really great.

One minor thing, I think this should be three dots icon (as it shows menu):

dots

Too bad that it's not possible to search for settings from main screen (only from subscreens) but I think this is too much work at this point. I'd also want to merge it ASAP.

File: app/k9mail/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -314,7 +314,12 @@
 
         <activity
             android:name=".ui.settings.AboutActivity"
-            android:label="@string/about_action" />
+            android:label="@string/about_action">
+            <intent-filter>
+                <action android:name="com.fsck.k9.OpenAbout" />
```
Comment (wiktor-k) [inline app/k9mail/src/main/AndroidManifest.xml]: :thinking: I think action names follow the same naming scheme as categories that is last part should be `UPPER_CASE`. E.g. `android.intent.action.MAIN`

Comment (williamvds) [reply app/k9mail/src/main/AndroidManifest.xml]: I saw that K9-specific ones tended to use camel case for actions (though it's a bit of a mixture), eg 
https://github.com/k9mail/k-9/blob/08172c42e1d569ac139f0f20667765756233e0ae/app/k9mail/src/main/AndroidManifest.xml#L346-L356
Is upper cased preferred?

**Comment by cketti** [comment]
While it's nice to be able to check out a proposed design on an actual device, maybe this should have been a GitHub issue to reach rough consensus on a design before starting an implementation.

Here are my thoughts on the settings screens. Please note that I haven't looked at the code at all.
- I liked the separation between general settings and account settings. I think folding general settings into the main list only clutters it up. I'm not fundamentally opposed to combining them. But there are a couple of issues that have to be addressed in order to make this a viable option:
  - Having search only available in subcategories is not an option.
  - Having search at the top level raises interesting problems for search. Because now we have a combined search for general settings and account settings (I'm aware that we currently don't have search for account settings, but it's definitely something we want to have in the future). One question is which account to open when a setting was found in account settings and the user clicks on the search result.
  - There should be better separation of list items. There could be a "General settings" heading like we already have for accounts. "About" definitely doesn't belong in there and should go to the bottom. Maybe under a "Miscellaneous" heading.
- I think using the account color for the envelope icon adds way too much color to this screen. Especially for people with lots of accounts.
- The floating action button is supposed to be for "important or common actions" in an app. I don't think adding an account qualifies for that. Again, that's a change I'd rather not include.
- Please don't add unrelated new features or placeholders for a new feature in a pull request (the settings import/export menu; doesn't work for me; I haven't checked the code to see if it's a bug or a placeholder)
- I think having the account dropdown be a separate line in the toolbar doesn't look great. My idea (sorry for not elaborating on this earlier) was to replace the title with the dropdown. In the collapsed state it should look like what our current title looks like ("Account settings" with the email address below). When there's only one account no dropdown should be displayed.

In summary: I don't see the benefit of folding general settings into the main settings screen right now. But if you want to pursue it, please make sure to address all points mentioned above.
I think this PR should be reduced to adding an account switcher to account settings. If you want to do more, please create additional PRs (ideally, one change at a time, not a "let's change everything" PR).

**Comment by williamvds** [comment]
Thanks for the explanation of your concerns
For now, I've simplified this pull request into just adding the spinner into the toolbar, as  requested

accountsettingsredesign3

Apologies for how this turned into a such a large request, I'll avoid doing that in the future

Adding search to the main activity, and fixing import/export was a TODO - I probably should have requested the 'dont merge' tag

> If search can open an account setting...

Maybe we could choose the account in this order
  - if user searched from account settings, use the currently selected account
  - if user searched and opened account settings during this session, use the last account they were modifying?
  - open using default account, relying on the user picking from the account spinner

> Separating setting list items

I agree with all your suggestions

> Remove account color in envelope icon

I see how the colors could get out of had, that's fine

> Remove FAB

I think I just kept it in from the initial redesign proposal, it now makes more sense to remove it

> Merge account dropdown into header

Done, activity title is embedded into the spinner's view
Dropdown is disabled and arrow hidden when there's only one account

**Comment by cketti** [comment]
Sorry for not getting to this earlier. 

I like the look of the account switcher :+1: Unfortunately, there are a couple of issues with the implementation. Detailed feedback follows in a bit.

File: app/ui/src/main/java/com/fsck/k9/ui/settings/account/AccountSettingsActivity.kt [hunk 3]
```diff
@@ -38,9 +52,31 @@ class AccountSettingsActivity : K9Activity(), OnPreferenceStartScreenCallback {
         loadAccount()
     }
 
+    override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
+        val uuid = accountSpinner.selection.uuid
+
+        if (uuid == accountUuid)
+            return;
+
+        start(this, uuid)
```
Comment (cketti) [inline app/ui/src/main/java/com/fsck/k9/ui/settings/account/AccountSettingsActivity.kt]: In my opinion starting a new Activity doesn't make for a nice user experience. Instead we could clear the fragment stack and add a fragment for the new account. 

However, being able to switch accounts requires quite a few additional changes. So this should probably be done in a separate PR.

File: app/ui/src/main/java/com/fsck/k9/ui/settings/account/AccountSettingsActivity.kt [hunk 4]
```diff
@@ -38,9 +52,31 @@ class AccountSettingsActivity : K9Activity(), OnPreferenceStartScreenCallback {
...
+
     private fun initializeActionBar() {
         val actionBar = supportActionBar ?: throw RuntimeException("getSupportActionBar() == null")
         actionBar.setDisplayHomeAsUpEnabled(true)
+        actionBar.setDisplayShowTitleEnabled(false)
+
+        accountSpinner.title = title
+
+        accountSpinner.onItemSelectedListener = this
+        val prefs = Preferences.getPreferences(this)
```
[the remark this answers is not in the document]
Comment (williamvds) [reply app/ui/src/main/java/com/fsck/k9/ui/settings/account/AccountSettingsActivity.kt]: Done, required handling some race conditions in `AccountSelectionSpinner`

File: app/ui/src/main/java/com/fsck/k9/ui/settings/account/AccountSettingsActivity.kt (hunk 3 above)
Comment (williamvds) [reply app/ui/src/main/java/com/fsck/k9/ui/settings/account/AccountSettingsActivity.kt]: OK, leaving as-is for this PR

File: app/ui/src/main/java/com/fsck/k9/ui/settings/account/AccountSettingsActivity.kt [hunk 5]
```diff
@@ -38,9 +52,31 @@ class AccountSettingsActivity : K9Activity(), OnPreferenceStartScreenCallback {
...
+        val viewModel: SettingsViewModel by viewModel()
+        viewModel.accounts.observeNotNull(this) {
+            accountSpinner.loadAccounts(prefs.accounts)
```
[the remark this answers is not in the document]
Comment (williamvds) [reply app/ui/src/main/java/com/fsck/k9/ui/settings/account/AccountSettingsActivity.kt]: The odd behaviour this was intended to resolve is now properly handled by `AccountSelectionSpinner`

File: app/ui/src/main/java/com/fsck/k9/ui/settings/account/AccountSettingsActivity.kt [hunk 6]
```diff
@@ -114,3 +157,69 @@ class AccountSettingsActivity : K9Activity(), OnPreferenceStartScreenCallback {
...
+        var title: CharSequence = ""
+
+        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
+            val account = getItem(position)
+
+            val view = convertView
+                ?: LayoutInflater.from(context).inflate(R.layout.account_spinner_item, parent, false)
+
+            return view.apply {
+                setPadding(0, paddingTop, paddingRight, paddingBottom)
```
[the remark this answers is not in the document]
Comment (williamvds) [reply app/ui/src/main/java/com/fsck/k9/ui/settings/account/AccountSettingsActivity.kt]: IIRC it was to avoid having to duplicate the layout. I've added another layout for the drop-down and removed the uses of `setPadding`
No idea why I did it for both the item and drop-down

File: app/ui/src/main/res/layout/account_spinner_item.xml [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    android:paddingRight="16dp">
+
+    <TextView
+        android:id="@+id/name"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:ellipsize="end"
+        android:singleLine="true"
+        android:textAppearance="?android:attr/textAppearanceListItem"
+        android:autoSizeTextType="uniform"
```
[the remark this answers is not in the document]
Comment (williamvds) [reply app/ui/src/main/res/layout/account_spinner_item.xml]: I believe it was an attempt to make landscape layout not mess up the positioning of the spinner, but as it turns out my issue was having set vertical padding on the spinner
It now achieves nothing, so I've removed it

File: app/ui/src/main/res/layout/activity_account_settings.xml [hunk 8]
```diff
@@ -4,7 +4,17 @@
...
-    <include layout="@layout/toolbar" />
+    <android.support.v7.widget.Toolbar
+                android:id="@+id/toolbar"
+                android:layout_width="match_parent"
+                android:layout_height="?attr/actionBarSize"
+                android:background="?attr/colorPrimary"
+                android:elevation="4dp">
+        <com.fsck.k9.ui.settings.account.AccountSelectionSpinner
+                    android:id="@+id/accountSpinner"
+                    android:layout_width="fill_parent"
```
[the remark this answers is not in the document]
Comment (williamvds) [reply app/ui/src/main/res/layout/activity_account_settings.xml]: I've changed it to `wrap_content` as you requested
The original reasoning was to have the drop-down width be fixed to toolbar's width rather than the content. Personal preference, I suppose

**Comment by williamvds** [comment]
I've applied your requested changes @cketti, thanks a bunch for your help!

**Comment by cketti** [comment]
Thanks!
